kernel/userspace.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <string.h>
+#include <misc/printk.h>
+#include <kernel_structs.h>
+#include <sys_io.h>
+
+/**
+ * Kernel object validation function
+ *
+ * Retrieve metadata for a kernel object. This function is implemented in
+ * the gperf script footer, see gen_kobject_list.py
+ *
+ * @param obj Address of kernel object to get metadata
+ * @return Kernel object's metadata, or NULL if the parameter wasn't the
+ * memory address of a kernel object
+ */
+__weak _GENERIC_SECTION(.kobject_data.text)
+struct _k_object *_k_object_find(void *obj)
+{
+	/* This is the __weak stub implementation, the version created by
+	 * gperf will replace it in build phase 2
+	 */
+	ARG_UNUSED(obj);
+	return NULL;
+}
+
+#ifdef CONFIG_PRINTK
+const char *otype_to_str(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return "k_alert";
+	case K_OBJ_DELAYED_WORK:
+		return "k_delayed_work";
+	case K_OBJ_MEM_SLAB:
+		return "k_mem_slab";
+	case K_OBJ_MSGQ:
+		return "k_msgq";
+	case K_OBJ_MUTEX:
+		return "k_mutex";
+	case K_OBJ_PIPE:
+		return "k_pipe";
+	case K_OBJ_SEM:
+		return "k_sem";
+	case K_OBJ_STACK:
+		return "k_stack";
+	case K_OBJ_THREAD:
+		return "k_thread";
+	case K_OBJ_TIMER:
+		return "k_timer";
+	case K_OBJ_WORK:
+		return "k_work";
+	case K_OBJ_WORK_Q:
+		return "k_work_q";
+	default:
+		return "?";
+	}
+}
+#endif
+
+/* Stub functions, to be filled in forthcoming patch sets */
+
+static void set_thread_perms(struct _k_object *ko, struct k_thread *thread)
+{
+	ARG_UNUSED(ko);
+	ARG_UNUSED(thread);
+
+	/* STUB */
+}
+
+
+static int test_thread_perms(struct _k_object *ko)
+{
+	ARG_UNUSED(ko);
+
+	/* STUB */
+
+	return 1;
+}
+
+static int _is_thread_user(void)
+{
+	/* STUB */
+
+	return 0;
+}
+
+void k_object_grant_access(void *object, struct k_thread *thread)
+{
+	struct _k_object *ko = _k_object_find(object);
+
+	if (!ko) {
+		if (_is_thread_user()) {
+			printk("granting access	to non-existent kernel object %p\n",
+			       object);
+			k_oops();
+		} else {
+			/* Supervisor threads may at times instantiate objects
+			 * that ignore rules on where they can live. Such
+			 * objects won't ever be usable from userspace, but
+			 * we shouldn't explode.
+			 */
+			return;
+		}
+	}
+
+	/* userspace can't grant access to objects unless it already has
+	 * access to that object
+	 */
+	if (_is_thread_user() && !test_thread_perms(ko)) {
+#ifdef CONFIG_PRINTK
```

Comment:
Not sure if "PRINTK" is really the right predicate for "I want runtime error logging out of the memory protection layer".

Can't this be handled by the existing SYS_LOG_*() framework?

Reply:
these ifdefs are only here to prevent build errors from use of otype_to_str() when CONFIG_PRINTK is turned off (and hence the function doesn't exist)